lib/src/file_picker_io.dart

```diff
@@ -16,6 +16,8 @@ const EventChannel _eventChannel =
 
 /// An implementation of [FilePicker] that uses method channels.
 class FilePickerIO extends FilePicker {
+  static void registerWith() {}
```

Comment:
Don't you have to set `FilePicker.platform` to an instance of `FilePickerIO()` in `registerWith()` ?
(and the other implementations in the other registerWith() functions)

You might also need to double check the one for web.

Then you could get rid of the `filePickerWithFFI()` stub and fix a bunch of issues relating to using the `Platform` class on the web.

The following issues are caused by that bug: 
https://github.com/miguelpruivo/flutter_file_picker/issues/1484
https://github.com/miguelpruivo/flutter_file_picker/issues/1483
https://github.com/miguelpruivo/flutter_file_picker/issues/1343

If you could get that fixed in this PR also, that would be great, since it has been a bit of a pain to deal with (and I didn't have time to look into fixing it myself yet)

Reply:
Thanks for your review and additional clarification! My goal here was to provide minimal effort fix for the warning, hence this dummy implementation. 
I would try to make changes to address your comment soon, thanks.

Reply:
> Don't you have to set FilePicker.platform to an instance of FilePickerIO() in registerWith() ?

About this – as far as I understand, it is the best practice :) So yes, I guess we can move logic there and it is preferred option. I would try that.
But current variant works because all logic for platform implementation choosing is already in _setPlatform(), so we needed only dummy empty method to satisfy compiler.

Reply:
Well, I dived in this issue and I don't think that I'm happy with what I found :)

When platform implementation is added as `dartPluginClass`, this implementation must be exported from main library file, as per [dart_plugin_registrant template](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/flutter_plugins.dart#L621-L673)

In our case, it means that `lib/file_picker.dart` will look something like this for registrant to be able to resolve implementation classes:

```dart
library file_picker;

export './src/file_picker.dart';
export './src/platform_file.dart';
export './src/file_picker_result.dart';

// Added lines start from here

export './src/file_picker_macos.dart';
export './src/linux/file_picker_linux.dart';
export './src/file_picker_io.dart';
export './src/windows/file_picker_windows.dart';
```

Of course it will break in web build on last line when exporting windows implementation for very same reason (ffi types) which `filePickerWithFFI()` needed for. Basically, this abstraction just goes to another level, where, instead of stub implementation, stub export is needed.

I will investigate further, but at this point I'm not sure that some appropriate solution, without creation of federation plugins, is possible. 

Would love to hear another opinions.

Reply:
According to https://docs.flutter.dev/packages-and-plugins/developing-packages#dart-only-platform-implementations

you don't specifically have to export these? Exporting an implementation would only be needed if there is extra functionality exposed by the specific implementation (i.e. an extra method that only exists for that implementation)

Currently the plugin does not expose such "platform specific extra functionality". If that is something we want to provide, we should switch to providing sub-packages for each platform.

And if we absolutely need a conditional export, you could do the following:

1) In https://github.com/miguelpruivo/flutter_file_picker/blob/master/lib/src/windows/file_picker_windows.dart change the existing stub to return an instance of `FilePickerWindows`

```
FilePickerWindows filePickerWithFFI() => FilePickerWindows();
```

2) Use the existing stub from https://github.com/miguelpruivo/flutter_file_picker/blob/master/lib/src/windows/stub.dart
as the stub implementation for the conditional export.

Reply:
> you don't specifically have to export these? 

I've also read this part of docs and thought like that too. But there is implementation detail which is not throughly described in documentation: framework generates `dart_plugin_registrant.dart` file where it calls `registerWith()` methods of concrete implementation. After `pubspec.yaml` modifications file looks like this:

```dart
//
// Generated file. Do not edit.
// This file is generated from template in file `flutter_tools/lib/src/flutter_plugins.dart`.
//

// @dart = 2.19

import 'dart:io'; // flutter_ignore: dart_io_import.
import 'package:file_picker/file_picker.dart';
import 'package:file_picker/file_picker.dart';
import 'package:file_picker/file_picker.dart';
import 'package:file_picker/file_picker.dart';
import 'package:file_picker/file_picker.dart';

@pragma('vm:entry-point')
class _PluginRegistrant {

  @pragma('vm:entry-point')
  static void register() {
    if (Platform.isAndroid) {
      try {
        FilePickerIO.registerWith();
      } catch (err) {
        print(
          '`file_picker` threw an error: $err. '
          'The app may not function as expected until you remove this plugin from pubspec.yaml'
        );
      }

    } else if (Platform.isIOS) {
      try {
        FilePickerIO.registerWith();
      } catch (err) {
        print(
          '`file_picker` threw an error: $err. '
          'The app may not function as expected until you remove this plugin from pubspec.yaml'
        );
      }

    } else if (Platform.isLinux) {
      try {
        FilePickerLinux.registerWith();
      } catch (err) {
        print(
          '`file_picker` threw an error: $err. '
          'The app may not function as expected until you remove this plugin from pubspec.yaml'
        );
      }

    } else if (Platform.isMacOS) {
      try {
        FilePickerMacOS.registerWith();
      } catch (err) {
        print(
          '`file_picker` threw an error: $err. '
          'The app may not function as expected until you remove this plugin from pubspec.yaml'
        );
      }

    } else if (Platform.isWindows) {
      try {
        FilePickerWindows.registerWith();
      } catch (err) {
        print(
          '`file_picker` threw an error: $err. '
          'The app may not function as expected until you remove this plugin from pubspec.yaml'
        );
      }

    }
  }
}
```

As you can see, it only imports `package:file_picker/file_picker.dart`. And if I'm not wrong, there is no possibility to change this behaviour.


As for conditional import – I considered these options too. But I believe that some kind of better implementation is possible :) Thing is, while breaking web build, those exported members are never really used there. That's why I think that stub implementation is a little bit overkill and, maybe, just export a completely empty stub file is more appropriate.

But, again, I will investigate further and come back with some solution shortly.

And, of course, thanks for your guidance and feedback!

Reply:
In the short term, you will indeed need to either do a conditional import layer (`path_provider_windows` and `path_provider_linux` have two different variants of that approach) to make the implementation class web-safe, or use the stub `defaultClass` approach.

Reply:
@stuartmorgan thanks!